### DISCUSSION TITLE
Is there a mechanism for a BIO write to update the SSL state?

### DISCUSSION THREAD

**Discussion by RealTrisT** [openssl/openssl#23230] [Q&A] [answered]
Using a memory BIO as an example:
```c
ssl = SSL_new(this->ctx);

ssl_input = BIO_new(BIO_s_mem());
ssl_output = BIO_new(BIO_s_mem());

SSL_set_bio(ssl, ssl_input, ssl_output);
```
Is there a way to make it so `BIO_write(ssl_input, buff, len)` would automatically update the SSL state such that a subsequent `SSL_pending` would yield the amount of data processed from the contents of `buff`?

I have tested that calling `SSL_read(ssl, 0, 0)` after the `BIO_write` effectively does that update, however with the existence of BIO pairs and BIO chains, I figured there's probably a more correct way to do it.

**Comment by mattcaswell** [comment, marked as answer]
> I have tested that calling SSL_read(ssl, 0, 0) after the BIO_write effectively does that update, however with the existence of BIO pairs and BIO chains, I figured there's probably a more correct way to do it.

This is the correct way to do it.
